### ISSUE TITLE
Get error when use specialization 

### ISSUE THREAD

**Issue by osanimmnt** [archimatetool/archi-modelrepository-plugin#177]
### Steps to Reproduce the Behaviour
1.
2.
3.

**Comment by Phillipus** [comment]
Hi, could you provide some details about the issue?

**Comment by osanimmnt** [comment]
> Hi, could you provide some details about the issue?

Sure. details added

**Comment by Phillipus** [comment]
@jbsarrodie Can you say what the main issues are regarding coArchi and Specializations? Are there any quick fixes we can apply to coArchi to handle them?

**Comment by Phillipus** [comment]
@osanimmnt Does the error occur on commit, or when you push and then refresh on another machine?

**Comment by jbsarrodie** [comment]
Hi,

First, as written [on the forum](https://forum.archimatetool.com/index.php?topic=1049.msg5848#msg5848):
>For those interrested in the new specialization mecanism in Archi 4.9: note that there is no support planned for it in coArchi (this will have to wait for coArchi 2 which will certainly take several months to be out). This doesn't mean you can't use both, but in case of conflicts there will be no mean to decide which "version" to keep, and thus this will most certainly block the merge process and thus the publication of changes on the remote server.
>You've been warned ;-)

> @jbsarrodie Can you say what the main issues are regarding coArchi and Specializations? Are there any quick fixes we can apply to coArchi to handle them?

I'll look at it. I think it should be quite easy to make it work (slight changes in the underlying GRAFICO engine), except for conflict handling which would require changes on the UI itself (to show concept's specialization and to expose specialization's definition)

**Comment by Phillipus** [comment]
@jbsarrodie Thanks. If you can give me the analysis and what's needed, I'll implement it.

**Comment by osanimmnt** [comment]
> r when you push and then refresh on another machine?

nop, commit is ok. it it when someone try to refresh model in other computer

**Comment by osanimmnt** [comment]
I create this Specialization
![image](https://user-images.githubusercontent.com/84808315/134350281-a47b2f1c-85e2-4bc6-8727-848c30bd9af4.png)
 and i assigned to a node:
![image](https://user-images.githubusercontent.com/84808315/134350461-5b3f6e99-f814-4bce-af9e-34eef4dd4c16.png)

commit ok.

and when someone else refresh model get this:
![image](https://user-images.githubusercontent.com/84808315/134350530-d937c226-e8a3-4e7c-895b-064caa6d1026.png)

**Comment by osanimmnt** [comment]
> @osanimmnt Thanks for the info.
> 
> As Archi 4.9 is still in beta and we haven't yet implemented support for Specializations in coArchi it's best not to use Specializations in models shared with coArchi.

:'( so bad, because it is a great functionallity to create very good views the users love!

**Comment by Phillipus** [comment]
@jbsarrodie Exporting to Grafico the reference to the Profile is using an absolute file href:

```xml
<archimate:BusinessRole
    xmlns:archimate="http://www.archimatetool.com/archimate"
    name="Business Role"
    id="id-e3873a0fb2db49d7bd8b8d42f46d6326">
  <profiles
      href="file:/C:/Users/Phillipus/AppData/Roaming/Archi4/model-repository/archisurance1/model/folder.xml#id-5348a711271e41f0b2d77c791046fde6"/>
</archimate:BusinessRole>
```

But when referencing from an object from a View it's not using `file:/`:

```xml
    <archimateElement
        xsi:type="archimate:BusinessActor"
        href="BusinessActor_id-fe9cddb8e22d420ba22e2c475520b97a.xml#id-fe9cddb8e22d420ba22e2c475520b97a"/>
```

I can't remember why or how this happens?

**Comment by Phillipus** [comment]
I think the place to look is in `GraficoExporter#createAndSaveResource`:

```java
URI key = file.getName().equals(IGraficoConstants.FOLDER_XML) ?
        URI.createFileURI(file.getAbsolutePath()) : URI.createFileURI(file.getName());

URI.createFileURI(file.getAbsolutePath()) : URI.createFileURI(file.getName());
URI value = URI.createFileURI(file.getAbsolutePath());
fResourceSet.getURIConverter().getURIMap().put(key, value);
```

`folder.xml` files are stored with absolute file paths, and other resources with file name. Therefore references to profiles in the main `folder.xml` file are absolute.

**Comment by Phillipus** [comment]
Adding this line solves the problem:

```java
if(object instanceof IArchimateModel) {
   key = URI.createFileURI(file.getName());
}
```

```xml
<archimate:BusinessRole
    xmlns:archimate="http://www.archimatetool.com/archimate"
    name="Business Role"
    id="id-e3873a0fb2db49d7bd8b8d42f46d6326">
  <profiles
      href="folder.xml#id-5348a711271e41f0b2d77c791046fde6"/>
</archimate:BusinessRole>
```

But don't know yet whether `href="folder.xml#id-5348a711271e41f0b2d77c791046fde6"` will resolve when importing...

**Comment by jbsarrodie** [comment]
> But don't know yet whether href=\"folder.xml#id-5348a711271e41f0b2d77c791046fde6\" will resolve when importing...

I think it wont because of the "folder.xml" in the fragment. There's some slight changes needed. Nothing complicated, I just need some time to work on it :-)

**Comment by Phillipus** [comment]
> I think it wont because of the "folder.xml" in the fragment. There's some slight changes needed. Nothing complicated, I just need some time to work on it :-)

No it doesn't resolve. I'll wait until you get some time ;-)

**Comment by osanimmnt** [comment]
did you find the error?

**Comment by Phillipus** [comment]
> did you find the error?

Yes. As you can see from the previous comments, we're working on it.

**Comment by jbsarrodie** [comment]
> I'll wait until you get some time ;-)

Done in edb69c8ef046c72dd2b17933f65ec84e811e2aeb

**Comment by Phillipus** [comment]
Thanks!

One question regarding this in `resolveProxies()`:

```java
if(eObject instanceof IArchimateConcept) {
...
...
} else if(eObject instanceof IArchimateRelationship) {
```

As `eObject is an `IArchimateRelationship`  it will execute the first block as it inherits from `IArchimateConcept`. Maybe remove the `else`?

**Comment by Phillipus** [comment]
And `concept.getProfiles();` might return an empty list.

**Comment by jbsarrodie** [comment]
Ok, see this as a POC ;-)

I'll fix it completely tomorrow.

**Comment by Phillipus** [comment]
> Ok, see this as a POC ;-)

Aha! :-)

**Comment by jbsarrodie** [comment]
Should be good with 3ba562babe8753a5375a53f46bdba700c84ac042 (I've force pushed an amended commit)

**Comment by Phillipus** [comment]
My OCD is playing up again...

`import com.archimatetool.model.IProfiles;` <------------------ 😟

**Comment by Phillipus** [comment]
> Should be good now ;-)

Phew, I can stop shaking now... 😄

**Comment by jbsarrodie** [comment]
> I downloaded and installes and new version of archimate and new version of plug-ins but the error pesist.

Could you double check the version of Archi and your plugins and test with a new model (not one containing specialization but exported using an old version of the plugin).

**Comment by osanimmnt** [comment]
the archi version:

![image](https://user-images.githubusercontent.com/84808315/138344916-0918b647-d46d-436b-9e08-d800150477e7.png)

plug-in versions:
![image](https://user-images.githubusercontent.com/84808315/138344897-b9e18fae-2612-46e2-88ac-a9e461d86285.png)


yes, I tried everything without specialization and it worked properly. but, when I added some especialization I publish and all good, but, when I tried to refresh model from other PC I get that error

**Comment by osanimmnt** [comment]
listo
https://github.com/osanimmnt/TestArchiSpecialization
